src/libraries/System.Private.CoreLib/src/System/Buffers/SharedArrayPool.cs

```diff
@@ -18,32 +19,49 @@ namespace System.Buffers
     /// checking its processor number, because multiple threads could interleave on the same core, and because
     /// a thread is allowed to check other core's buckets if its core's bucket is empty/full.
     /// </remarks>
-    internal sealed partial class TlsOverPerCoreLockedStacksArrayPool<T> : ArrayPool<T>
+    internal sealed partial class SharedArrayPool<T> : ArrayPool<T>
     {
         /// <summary>The number of buckets (array sizes) in the pool, one for each array length, starting from length 16.</summary>
         private const int NumBuckets = 27; // Utilities.SelectBucketIndex(1024 * 1024 * 1024 + 1)
-        /// <summary>Maximum number of per-core stacks to use per array size.</summary>
-        private const int MaxPerCorePerArraySizeStacks = 64; // selected to avoid needing to worry about processor groups
-        /// <summary>The maximum number of buffers to store in a bucket's global queue.</summary>
-        private const int MaxBuffersPerArraySizePerCore = 8;
 
         /// <summary>A per-thread array of arrays, to cache one array per array size per thread.</summary>
         [ThreadStatic]
         private static ThreadLocalArray[]? t_tlsBuckets;
         /// <summary>Used to keep track of all thread local buckets for trimming if needed.</summary>
         private readonly ConditionalWeakTable<ThreadLocalArray[], object?> _allTlsBuckets = new ConditionalWeakTable<ThreadLocalArray[], object?>();
         /// <summary>
-        /// An array of per-core array stacks. The slots are lazily initialized to avoid creating
+        /// An array of per-core partitions. The slots are lazily initialized to avoid creating
         /// lots of overhead for unused array sizes.
         /// </summary>
-        private readonly PerCoreLockedStacks?[] _buckets = new PerCoreLockedStacks[NumBuckets];
+        private readonly Partitions?[] _buckets = new Partitions[NumBuckets];
         /// <summary>Whether the callback to trim arrays in response to memory pressure has been created.</summary>
         private int _trimCallbackCreated;
 
-        /// <summary>Allocate a new PerCoreLockedStacks and try to store it into the <see cref="_buckets"/> array.</summary>
-        private PerCoreLockedStacks CreatePerCoreLockedStacks(int bucketIndex)
+        /// <summary>Gets the maximum number of partitions to shard arrays into.</summary>
+        /// <remarks>Defaults to int.MaxValue.  Whatever value is returned will end up being clamped to <see cref="Environment.ProcessorCount"/>.</remarks>
+        private static int GetPartitionCount()
         {
-            var inst = new PerCoreLockedStacks();
+            string? value = Environment.GetEnvironmentVariable("System.Buffers.SharedArrayPool.MaxPartitionCount");
+            int partitionCount = value is not null && int.TryParse(value, CultureInfo.InvariantCulture, out int result) && result > 0 ?
```

Comment:
It looks dangerous to call globalization from here. There are paths through globalization that are going to call back into array pool. It is hard to reason about all possible paths we can take and the problems that this can cause.

Reply:
`Environment.GetEnvironmentVariable` has similar potential problem. Interesting things will happen if you set this env variable is to a string that is longer than 128 and we hit the `EnsureCapacity` path through `GetEnvironmentVariableCore` on Windows.

Reply:
Ah, I addressed your first comment.  I didn't address the second.  Will follow-up on that one.